<a href="https://colab.research.google.com/github/Carolinsrainbow/UC_Seguridad/blob/main/Tarea_1_HerreraAzolasCarolina.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tarea: OTP en la práctica

El objetivo de esta tarea es que extienda el protocolo criptográfico OTP para que pueda funcionar correctamente cuando la clave es más corta que el mensaje, la cual es la situación usual en la práctica.

Como en el turorial sobre OTP, nos vamos a restringir a un conjunto de símbolos que consiste de las 26 letras del alfabeto Inglés en mayúscula y el símbolo espacio ``' '``, pero extendido con el salto de línea ``\n`` y algunos signos de puntuación. Para esto consideramos el siguiente diccionario que codifica cada uno de estos 32 símbolos como un número entero.

In [1]:
alfabeto = { 'A': 0, 'B': 1, 'C': 2, 'D': 3, 'E': 4, 'F': 5, 'G': 6, 'H': 7, 'I': 8, 'J': 9,
             'K': 10, 'L': 11, 'M': 12, 'N': 13, 'O': 14, 'P': 15, 'Q': 16, 'R': 17, 'S': 18,
             'T': 19, 'U': 20, 'V': 21, 'W': 22, 'X': 23, 'Y': 24, 'Z': 25, ' ': 26, '\n' : 27,
             ',': 28, '.': 29, '!': 30, ';': 31}

A continuación tenemos la función ``_EncOTP`` que permite cifrar un texto plano ``mensaje`` utilizando ``clave``, bajo el supuesto de que ambos tienen el mismo largo. Además, tenemos la función ``_DecOTP`` que permite descifrar un texto ``cifrado`` utilizando ``clave``, bajo el mismo supuesto de que ambos tienen el mismo largo. Ambas funciones, junto con una serie de procedimientos auxiliares necesarios para implementarlas, fueron explicadas en el tutorial sobre OTP. El único cambio que hemos realizado aquí es considerar un alfabeto con más símbolos. Su primera tarea es estudiar y entender en detalle cada una de las siguientes funciones.

In [2]:
def representar_texto_lista(m):
    lista = []
    for i in range(len(m)):
        lista.append(int(alfabeto[m[i]]))
    return lista


def simbolo(v):
    for s in alfabeto:
        if alfabeto[s] == v:
            return(s)


def representar_lista_texto(lista):
    mensaje = ""
    for i in range(len(lista)):
        mensaje += simbolo(lista[i])
    return mensaje


def _EncOTP(clave, mensaje):
    clave_lista = representar_texto_lista(clave)
    mensaje_lista = representar_texto_lista(mensaje)
    cifrado_lista = []
    for i in range(len(clave_lista)):
          cifrado_lista.append((mensaje_lista[i] + clave_lista[i]) % len(alfabeto))
    return representar_lista_texto(cifrado_lista)


def _DecOTP(clave, cifrado):
    clave_lista = representar_texto_lista(clave)
    cifrado_lista = representar_texto_lista(cifrado)
    mensaje_lista = []
    for i in range(len(clave_lista)):
          mensaje_lista.append((cifrado_lista[i] - clave_lista[i]) % len(alfabeto))
    return representar_lista_texto(mensaje_lista)

Nótese que utilizamos el símbolo ``_`` en el nombre de la función ``_EncOTP`` para indicar que esta función sólo es de uso interno, y no debería ser invocada por un usuario que quiera cifrar un mensaje. La razón para esto es que esta función está diseñada para funcionar con mensajes y claves del mismo largo, y la función ``EncOTP`` que usted va a diseñar en esta tarea tiene que funcionar en el caso general en el que este supuesto no es cierto. Por la misma razón, utilizamos la notación ``_DecOTP``, y usted va a tener que construir ``DecOTP``.

A continuación mostramos un ejemplo donde ciframos y desciframos un mensaje con las funciones anteriores.

In [3]:
mensaje = "JUZGADOS"
cifrado = _EncOTP("XHTODERF", mensaje)
descifrado = _DecOTP("XHTODERF", cifrado)
print(f"{mensaje}\n{cifrado}\n{descifrado}")

JUZGADOS
A
MUDH;X
JUZGADOS


Como esperábamos, en el ejemplo anterior el mensaje ``JUZGADOS`` y la clave ``XHTODERF`` tienen el mismo largo.

Para poder utilizar OTP cuando la clave es más corta que el mensaje, vamos a utilizar la clave como la semilla en un generador de números aleatorios, el cual vamos a utilizar para generar una nueva clave tan larga como el mensaje. Para implementar este enfoque, vamos a utilizar la librería ``random`` de Python, y la función ``randint``.

In [4]:
import random

for i in range(10):
    print(random.randint(0, 1))

1
1
1
1
0
1
1
1
1
0


La llamada ``randint(a, b)`` genera un número entero aleatorio ``x`` tal que ``a <= x <= b``. La clase ``int`` de Python puede manejar números enteros de largo arbitrario, y de la misma forma ``randint`` puede generar números enteros aleatorios de largo arbitrario.

In [5]:
for i in range(10):
    print(random.randint(-(10**100), 10**100))

4315409940237223174031948269695218207878291675540355277621952344620281613487436943843559345047808040
-3906431369340496602632439379037341667361871116327994263178025081418342881222105862778278125806167675
4693073538033581480756709903153156197850652412967575551476200758908368512737804638768171523768362374
7137258402407972444112519153149933309589155696007242266747775667271717619795257878492507687203536870
4874363119381136563685367560953485976927575454558647789405123381223458313969456204310354246710346458
504216192693586772169409094562087960909027131704376187796462301929026388558748582752347483918238076
8853846777655711872571062641596322336809251588203182760238034834172126783179930021530358352167046505
2954588092279684408624314123497548136744145433889636560499532276293188064999105199379389907331363669
3398588856479163394523512384375770708227794227389247974725256955193962638157371378463700877561752011
-568723664655856071473428346714948550618981006302751978388410417842947554420486600512805564

Los números aleatorios generados por ``randint`` son contruidos a partir de un valor inicial o **semilla**. Si esta semilla no es dada de manera explicita, entonces es definida por Python y nos encontramos con valores aleatorios distintos cada vez que llamamos a la función ``randint``.

In [6]:
for i in range(10):
    print(random.randint(-(10**100), 10**100))

-27893982712957348778761147641398924784836650038395538819930885117517239825328404145352289698545100
3847301690060905825915459015682767014871816170449534466883651751435917182226040607664653588823230462
6401608871848682463206981431329834025857501561072787428184991791497483494303785990343713911183699358
-7902732653550101077605379268237628521665422401315945734912714027704786265301604671521316620987726664
-6934586674434379339580387655990477465992862068116733383488032770677939300916502062501450080457426018
5046492389632567148133942064444363387861025893034123206383830620208024766786421457511746039638534372
8860774184354597084246412518218030275947129289468998659993500438847464855814953783659547071418496170
-6419188695580848725251822452908330318187294953684384078517866255129291407063754813714752329069196733
4816048499125026894593571121788384493036681638603820356306047221693983373539407574595574459049571394
83752265747904994890625728169328125293737083966007153964921173044478689587696536488042137

Sin embargo, la semilla de ``randint`` puede ser definida a través de la función ``seed``.

In [7]:
random.seed(1000)

for i in range(10):
    print(random.randint(-(10**100), 10**100))

8711600494022973757927074957965738644324128741809009356649523315171551510748054244168768592042580044
-2964297909258499355352801805179716129421520159173882863947087952916278756606899041144295924180492060
-9963489872032333227341178705455085122118478833301134944649477678776566957866337532257054686524871222
9182807805484118137516965655759002276321254687919498390083009759262573565802661851132993689732144986
2500290660143724019404560076353983650719048893941755008330583525340080576650024668964166297662038878
6971199532614052867771130714762498547287437544297752978655067807933943030412760963712114940611591305
5740393701883547881209102691360641389727694170081387814204769307404558408798100963342531500957368135
2963004951926965706231848636313286172021054863995325009853706177347418396716738196505272356719557256
493874583105159510691314383979104638416987172271760647113291914565433057735545549788727516458330748
-42056943739058671565720281896275939850997324367523530744140643397123869940997001838922858

Si utilizamos la misma semilla, entonces siempre vamos a ver los mismos números aletarorios generados.

In [8]:
random.seed(1000)

for i in range(10):
    print(random.randint(-(10**100), 10**100))

8711600494022973757927074957965738644324128741809009356649523315171551510748054244168768592042580044
-2964297909258499355352801805179716129421520159173882863947087952916278756606899041144295924180492060
-9963489872032333227341178705455085122118478833301134944649477678776566957866337532257054686524871222
9182807805484118137516965655759002276321254687919498390083009759262573565802661851132993689732144986
2500290660143724019404560076353983650719048893941755008330583525340080576650024668964166297662038878
6971199532614052867771130714762498547287437544297752978655067807933943030412760963712114940611591305
5740393701883547881209102691360641389727694170081387814204769307404558408798100963342531500957368135
2963004951926965706231848636313286172021054863995325009853706177347418396716738196505272356719557256
493874583105159510691314383979104638416987172271760647113291914565433057735545549788727516458330748
-42056943739058671565720281896275939850997324367523530744140643397123869940997001838922858

En esta tarea, dado un texto plano ``mensaje`` y una clave secreta ``clave``, usted debe utilizar la función ``seed`` para indicar que ``clave`` es la semilla que será utilizada por ``randint`` para generar números aleatorios que serán utilizados para construir una clave tan larga como ``mensaje``, de manera tal de poder utilizar ``_EncOTP`` para poder cifrar ``mensaje``.

Como primera tarea, debe implementar la siguiente función que considera ``clave`` como un número en base ``32``, y lo transforma en su representacional decimal.

In [19]:
def representar_numero(clave):
    numero = 0
    base = 32
    for caracter in clave:
        numero = numero * base + alfabeto[caracter]
    return numero

In [20]:
representar_numero(" BA.")

853021

Por ejemplo, la representación de ``" BA."`` como número decimal es ``853021`` puesto de acuerdo al diccionario ``alfabeto`` el espacio representa al número ``26``, el símbolo ``B`` representa al número ``1``, el símbolo ``A`` representa al número ``0``, y el símbolo ``.`` representa al número ``29``, por lo cual la representación decimal de ``" BA."`` es:

In [21]:
26*(32**3) + 1*(32**2) + 0*(32**1) + 29*(32**0)

853021

Dejamos a continuación algunos ejemplos con sus resultados para que pueda verificar que su implementación de la función ``representar_numero`` es correcta.

In [22]:
claves = ["XHTODERF", "A", "ESTE ES UN CURSO DE SEGURIDAD DE DATOS", " ESTE ES UN CURSO DE SEGURIDAD DE DATOS",
          "AESTE ES UN CURSO DE SEGURIDAD DE DATOS", "ESTE ES UN CURSO DE SEGURIDAD DE DATOSA", "AAAAA"]

for clave in claves:
    print(representar_numero(clave))

798442492453
0
224661477179998607071610158486595787614080674076652752338
41025822757193423572004241409336427492279391063092877086162
224661477179998607071610158486595787614080674076652752338
7189167269759955426291525071571065203650581570452888074816
0


En particular, en este caso usted debe obtener como resultado la siguiente lista de números:
```
798442492453
0
224661477179998607071610158486595787614080674076652752338
41025822757193423572004241409336427492279391063092877086162
224661477179998607071610158486595787614080674076652752338
7189167269759955426291525071571065203650581570452888074816
0
```

A continuación usted debe implementar la función ``representar_clave`` que puede entenderse como la función inversa de ``representar_numero``. Vale decir, ``representar_clave`` recibe como entrada un número natural ``numero`` y construye su representación como un número en base ``32`` según los símbolos del diccionario ``alfabeto``. Adicionalmente, la función ``representar_clave`` recibe el largo de la clave que debe construir.

In [23]:
def representar_clave(numero, largo):
    base = 32
    clave = ""
    inverso_alfabeto = {v: k for k, v in alfabeto.items()}
    for _ in range(largo):
        resto = numero % base
        clave = inverso_alfabeto[resto] + clave
        numero = numero // base
    return clave

Como un primer ejemplo del uso de esta función, considere lo siguiente.

In [24]:
n = representar_numero(" BA.")
c = representar_clave(n, 4)
print(n)
print(c)

853021
 BA.


Obtuvimos el resultado esperado, al igual que en el siguiente caso.

In [25]:
n = representar_numero("A")
c = representar_clave(n, 1)
print(n)
print(c)

0
A


¿Pero cómo podríamos obtener la clave ``AAAAA``? Para esto tenemos que especificar el largo apropiado.

In [26]:
n = representar_numero("A")
c = representar_clave(n, 5)
print(n)
print(c)

0
AAAAA


Los siguientes ejemplos muestras que la función ``representar_clave`` funciona como la inversa de la función ``representar_numero`` cuando es dado el largo apropiado. Usted debe utilizar estos ejemplos para verificar si su implementación de la función ``representar_clave`` es correcta.

In [27]:
claves = ["XHTODERF", "ESTE ES UN CURSO DE SEGURIDAD DE DATOS", " ESTE ES UN CURSO DE SEGURIDAD DE DATOS",
          "AESTE ES UN CURSO DE SEGURIDAD DE DATOS", "ESTE ES UN CURSO DE SEGURIDAD DE DATOSA", "UN MENSAJE ALEATORIO"]

for clave in claves:
    n = representar_numero(clave)
    c = representar_clave(n, len(clave))
    print(f"{clave}\n{c}\n")

XHTODERF
XHTODERF

ESTE ES UN CURSO DE SEGURIDAD DE DATOS
ESTE ES UN CURSO DE SEGURIDAD DE DATOS

 ESTE ES UN CURSO DE SEGURIDAD DE DATOS
 ESTE ES UN CURSO DE SEGURIDAD DE DATOS

AESTE ES UN CURSO DE SEGURIDAD DE DATOS
AESTE ES UN CURSO DE SEGURIDAD DE DATOS

ESTE ES UN CURSO DE SEGURIDAD DE DATOSA
ESTE ES UN CURSO DE SEGURIDAD DE DATOSA

UN MENSAJE ALEATORIO
UN MENSAJE ALEATORIO



En particular, en este caso usted debe obtener el siguiente resultado:
```
XHTODERF
XHTODERF

ESTE ES UN CURSO DE SEGURIDAD DE DATOS
ESTE ES UN CURSO DE SEGURIDAD DE DATOS

 ESTE ES UN CURSO DE SEGURIDAD DE DATOS
 ESTE ES UN CURSO DE SEGURIDAD DE DATOS

AESTE ES UN CURSO DE SEGURIDAD DE DATOS
AESTE ES UN CURSO DE SEGURIDAD DE DATOS

ESTE ES UN CURSO DE SEGURIDAD DE DATOSA
ESTE ES UN CURSO DE SEGURIDAD DE DATOSA

UN MENSAJE ALEATORIO
UN MENSAJE ALEATORIO
```

Ya tenemos los ingredientes para explicar como funciona la llamada ``EncOTP(clave, mensaje)``. Suponga que el largo de ``clave`` es ``c`` y el largo de ``mensaje`` es ``m``. Entonces ``EncOTP`` utiliza ``representar_numero`` para transformar ``clave`` en ``numero``, y utiliza ``numero`` como la semilla de ``randint``. Una vez hecho esto, utiliza ``randint`` para general números ``n_1``, ``n_2``, ..., cada uno entre ``0`` y ``32**c - 1``, los cuales son transformados en claves ``clave_1``, ``clave_2``, ..., cada una de largo ``c``. Este proceso continua hasta que la concatenación ``clave_1 clave_2 ... clave_t`` tenga un largo total mayor a igual a ``m``, en cuyo caso se deja los primeros ``m`` símbolos de ``clave_1 clave_2 ... clave_t`` como la clave a ser utilizada por OTP a través de la función ``_EncOTP``. La función ``DecOTP`` es definida de la misma forma.

In [29]:
def EncOTP(clave, mensaje):
    c = len(clave)
    m = len(mensaje)

    numero = representar_numero(clave)
    random.seed(numero)
    clave_total = ""
    while len(clave_total) < m:
        aleatorio = random.randint(0, 32**c - 1)
        clave_parcial = representar_clave(aleatorio, c)
        clave_total += clave_parcial
    clave_final = clave_total[:m]
    return _EncOTP(clave_final, mensaje)


def DecOTP(clave, mensaje):
    c = len(clave)
    m = len(mensaje)

    numero = representar_numero(clave)
    random.seed(numero)

    clave_total = ""
    while len(clave_total) < m:
        aleatorio = random.randint(0, 32**c - 1)
        clave_parcial = representar_clave(aleatorio, c)
        clave_total += clave_parcial

    clave_final = clave_total[:m]

    return _DecOTP(clave_final, mensaje)



A continuación dejamos una serie de ejemplos que ustede debe utilizar para verificar que sus implementaciones de las funciones ``EncOTP`` y ``DecOTP`` son correctas.

In [30]:
mensajes = ["HOLA", "ESTE ES UN PRIMER MENSAJE.", "ESTE ES UN SEGUNDO MENSAJE.",
            "AQUI TRATAMOS DE COLOCAR UNA FRASE MUCHO MAS LARGA!",
            "ESTE ES UN ULTIMO MENSAJE DE PRUEBA, EL CUAL ES MAS LARGO QUE LOS ANTERIORES."]

for mensaje in mensajes:
    cifrado = EncOTP("XHTODERF", mensaje)
    descifrado = DecOTP("XHTODERF", cifrado)
    print(f"{mensaje}\n{cifrado}\n{descifrado}\n--------------------")

HOLA
IEUO
HOLA
--------------------
ESTE ES UN PRIMER MENSAJE.
FI,S ETSTULLQMAQ ZWZTKYIR;
ESTE ES UN PRIMER MENSAJE.
--------------------
ESTE ES UN SEGUNDO MENSAJE.
FI,S ETSTULODKIZMNEBKFK;WG 
ESTE ES UN SEGUNDO MENSAJE.
--------------------
AQUI TRATAMOS DE COLOCAR UNA FRASE MUCHO MAS LARGA!
BG.W TSYSH.KR!XQDBYAU YQHWK
DFJDUWKEMZ.!KNHQO
GHSV,
AQUI TRATAMOS DE COLOCAR UNA FRASE MUCHO MAS LARGA!
--------------------
ESTE ES UN ULTIMO MENSAJE DE PRUEBA, EL CUAL ES MAS LARGO QUE LOS ANTERIORES.
FI,S ETSTULQKX,YXZWZTKYIR,A;DPJXGTQUS
BKSVHJOUYQYVQCOBGOORFPUTLHLROR
XVQJJNB

ESTE ES UN ULTIMO MENSAJE DE PRUEBA, EL CUAL ES MAS LARGO QUE LOS ANTERIORES.
--------------------


En particular, en este caso usted debe obtener el siguiente resultado:
```
HOLA
IEUO
HOLA
--------------------
ESTE ES UN PRIMER MENSAJE.
FI,S ETSTULLQMAQ ZWZTKYIR;
ESTE ES UN PRIMER MENSAJE.
--------------------
ESTE ES UN SEGUNDO MENSAJE.
FI,S ETSTULODKIZMNEBKFK;WG
ESTE ES UN SEGUNDO MENSAJE.
--------------------
AQUI TRATAMOS DE COLOCAR UNA FRASE MUCHO MAS LARGA!
BG.W TSYSH.KR!XQDBYAU YQHWK
DFJDUWKEMZ.!KNHQO
GHSV,
AQUI TRATAMOS DE COLOCAR UNA FRASE MUCHO MAS LARGA!
--------------------
ESTE ES UN ULTIMO MENSAJE DE PRUEBA, EL CUAL ES MAS LARGO QUE LOS ANTERIORES.
FI,S ETSTULQKX,YXZWZTKYIR,A;DPJXGTQUS
BKSVHJOUYQYVQCOBGOORFPUTLHLROR
XVQJJNB

ESTE ES UN ULTIMO MENSAJE DE PRUEBA, EL CUAL ES MAS LARGO QUE LOS ANTERIORES.
--------------------
```

Y como un último ejemplo, mostramos como se cifra utilizado OTP el poema "LOS HERALDOS NEGROS" de César Vallejo. Usted también debe utilizar este ejemplo para verificar que su implementación es correcta. En particular, debe revisar que el poema original se obtiene como resultado del proceso de cifrar y luego descifrar el poema con OTP.

In [31]:
mensaje = """LOS HERALDOS NEGROS

HAY GOLPES EN LA VIDA, TAN FUERTES... YO NO SE!
GOLPES COMO DEL ODIO DE DIOS; COMO SI ANTE ELLOS,
LA RESACA DE TODO LO SUFRIDO
SE EMPOZARA EN EL ALMA... YO NO SE!

SON POCOS, PERO SON... ABREN ZANJAS OSCURAS
EN EL ROSTRO MAS FIERO Y EN EL LOMO MAS FUERTE.
SERAN TAL VEZ LOS POTROS DE BARBAROS ATILAS;
O LOS HERALDOS NEGROS QUE NOS MANDA LA MUERTE.

SON LAS CAIDAS HONDAS DE LOS CRISTOS DEL ALMA,
DE ALGUNA FE ADORABLE QUE EL DESTINO BLASFEMA.
ESOS GOLPES SANGRIENTOS SON LAS CREPITACIONES
DE ALGUN PAN QUE EN LA PUERTA DEL HORNO SE NOS QUEMA.

Y EL HOMBRE... POBRE...POBRE! VUELVE LOS OJOS, COMO
CUANDO POR SOBRE EL HOMBRO NOS LLAMA UNA PALMADA;
VUELVE LOS OJOS LOCOS, Y TODO LO VIVIDO
SE EMPOZA, COMO CHARCO DE CULPA, EN LA MIRADA.

HAY GOLPES EN LA VIDA, TAN FUERTES... YO NO SE!"""

cifrado = EncOTP("XHTODERF", mensaje)
descifrado = DecOTP("XHTODERF", cifrado)
print(f"Mensaje:\n{mensaje}\n\n")
print(f"Cifrado:\n{cifrado}\n\n")
print(f"Mensaje descifrado:\n{descifrado}")

Mensaje:
LOS HERALDOS NEGROS

HAY GOLPES EN LA VIDA, TAN FUERTES... YO NO SE!
GOLPES COMO DEL ODIO DE DIOS; COMO SI ANTE ELLOS,
LA RESACA DE TODO LO SUFRIDO
SE EMPOZARA EN EL ALMA... YO NO SE!

SON POCOS, PERO SON... ABREN ZANJAS OSCURAS
EN EL ROSTRO MAS FIERO Y EN EL LOMO MAS FUERTE.
SERAN TAL VEZ LOS POTROS DE BARBAROS ATILAS;
O LOS HERALDOS NEGROS QUE NOS MANDA LA MUERTE.

SON LAS CAIDAS HONDAS DE LOS CRISTOS DEL ALMA,
DE ALGUNA FE ADORABLE QUE EL DESTINO BLASFEMA.
ESOS GOLPES SANGRIENTOS SON LAS CREPITACIONES
DE ALGUN PAN QUE EN LA PUERTA DEL HORNO SE NOS QUEMA.

Y EL HOMBRE... POBRE...POBRE! VUELVE LOS OJOS, COMO
CUANDO POR SOBRE EL HOMBRO NOS LLAMA UNA PALMADA;
VUELVE LOS OJOS LOCOS, Y TODO LO VIVIDO
SE EMPOZA, COMO CHARCO DE CULPA, EN LA MIRADA.

HAY GOLPES EN LA VIDA, TAN FUERTES... YO NO SE!


Cifrado:
ME
IHESYKK;OZRYS N,QB;YXHILGYEK.G;KDYRLYTBDYHQTQRJCZWFHF.UPT!TNHTJSCDZSTK,
JACWVZESDYQVWUI.JEPNM
H,TUKCTN C,OLYB,ZGPDH.JWW RFTXPST
AHEA;Q;
JYSX NCP UFEC;CB,NDOCTF
.B UYN
C!L!FIH